oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsScopeValidator.java

```diff
@@ -0,0 +1,35 @@
+package org.springframework.security.oauth2.server.authorization.authentication;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.springframework.security.oauth2.core.OAuth2AuthenticationException;
+import org.springframework.security.oauth2.core.OAuth2ErrorCodes;
+import org.springframework.util.CollectionUtils;
+
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.Set;
+
+public class OAuth2ClientCredentialsScopeValidator {
```

Comment:
@adamleantech There is an existing pattern for request validation that we want to follow. Please see `OAuth2AuthorizationCodeRequestAuthenticationProvider.setAuthenticationValidator(Consumer)` and the associated default implementation `OAuth2AuthorizationCodeRequestAuthenticationValidator`. Also see gh-891 and associated commits. Please adjust this PR to follow the same pattern. Thanks.

Reply:
@jgrandja Many thanks for pointing me in the right direction, I'll take a look at that

Reply:
@adamleantech At this point, we do not want to allow setting the validator through the Configurer. The reasons are outlined in gh-1003. Instead, the setter should be in the `AuthenticationProvider` and it would allow to either add to it or override it, for example, `OAuth2ClientCredentialsAuthenticationProvider.setAuthenticationValidator(Consumer<OAuth2ClientCredentialsAuthenticationContext> authenticationValidator)`.

Take a look at the [reference documentation](https://docs.spring.io/spring-authorization-server/docs/current/reference/html/protocol-endpoints.html#oauth2-authorization-endpoint-customizing-authorization-request-validation) for `OAuth2AuthorizationCodeRequestAuthenticationProvider`, as it demonstrates how to override.

Reply:
@jgrandja thanks for the help, hopefully the current state is in line with your thoughts.  If so I'll continue with tests and finalise the PR

Reply:
Looks good @adamleantech. Just a heads up that this enhancement is scheduled for `1.3.x`. We've already released `1.2.0-RC1`, so we're not introducing anymore public APIs for the `1.2.0` release.